### PR TITLE
NEW @W-21102140@ - Add PMD AST dump API for programmatic AST generation  

### PR DESCRIPTION
**Pull request by aruntyagiTutu**
@W-21102140@

Summary                                                                                                                                                                                                                                        
                                                                                                                                                                                                                                                 
  Adds new AST (Abstract Syntax Tree) dump functionality to the PMD engine, enabling programmatic generation of XML AST representations for Apex and Visualforce code without requiring PMD CLI.                                                 
                                                                                                                                                                                                                                                 
  What's New

  - New Public API: PmdEngine.generateAst() method for on-demand AST generation
  - Java Implementation: New ast-dump command in PmdWrapper using PMD's TreeExporter API
  - Type Safety: Full TypeScript type definitions (PmdAstDumpResults, GenerateAstOptions)

  Key Features

  - Supports Apex and Visualforce languages
  - Configurable file encoding (defaults to UTF-8)
  - Comprehensive error handling with structured error objects
  - Temporary working folder management with automatic cleanup
  - Memory-efficient validation (no unnecessary file loading)

  Changes Made

  1. Java layer (pmd-cpd-wrappers):
    - PmdAstDumper.java - Core AST generation logic
    - PmdAstDumpInputData.java - Input structure
    - PmdAstDumpResults.java - Output structure
    - Updated PmdWrapper.java - Added "ast-dump" command handler
  2. TypeScript layer:
    - Added generateAst() method to PmdEngine class
    - Exported PmdEngine for direct instantiation (needed for MCP provider)
    - Type exports for public API consumption
  3. Tests:
    - 18 Java unit tests (100% coverage)
    - 13 TypeScript integration tests
    - Edge cases: validation, encodings, error handling, empty files

  API Example

  const pmdEngine = new PmdEngine(config);
  const result = await pmdEngine.generateAst('apex', '/path/to/file.cls', {
    encoding: 'UTF-8',
    workingFolder: '/tmp/ast'
  });
  // result.ast contains XML AST or result.error contains error details

### CODE REVIEW

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java [hunk 1]
```diff
@@ -0,0 +1,122 @@
...
+     * @return Results containing either the AST (if successful) or error details (if failed)
+     */
+    public PmdAstDumpResults dump(PmdAstDumpInputData inputData) {
+        validateInputData(inputData);
+
+        PmdAstDumpResults results = new PmdAstDumpResults();
+        results.file = inputData.fileToDump;
+
+        try {
+            System.out.println("Generating AST for file '" + inputData.fileToDump + "' with language '" + inputData.language + "'");
```
Comment (namrata111f) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:36]: Do we need to log this instead?

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+            // Create TreeExportConfiguration
+            TreeExportConfiguration config = new TreeExportConfiguration();
+            config.setLanguage(language);
+            config.setFormat("xml"); // Always XML format for v1
+            config.setFile(filePath);
+
+            // Capture output to string
+            ByteArrayOutputStream baos = new ByteArrayOutputStream();
+            PrintStream ps = new PrintStream(baos, true, StandardCharsets.UTF_8);
+            PrintStream originalOut = System.out;
```
Comment (namrata111f) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:57]: The variable names look unclear.

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java [hunk 3]
```diff
@@ -0,0 +1,122 @@
...
+
+            try {
```
Comment (namrata111f) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:59]: There is no catch block corresponding to this try-block. Are we relying on outer catch block?

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java [hunk 4]
```diff
@@ -0,0 +1,122 @@
...
+                // Restore original System.out
+                System.setOut(originalOut);
+                ps.close();
+            }
+
+            System.out.println("Successfully generated AST for file '" + inputData.fileToDump + "'");
+
+        } catch (Exception e) {
+            // Store processing error
+            System.err.println("Error generating AST for file '" + inputData.fileToDump + "': " + e.getMessage());
```
Comment (namrata111f) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:80]: Any reason we are using System prints instead of logger here?

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java [hunk 5]
```diff
@@ -0,0 +1,122 @@
...
+    /**
+     * Validates file path and encoding without reading file content.
+     * This avoids loading large files into memory unnecessarily.
+     * PMD's TreeExporter will handle reading the file content.
+     */
+    private void validateFilePath(Path filePath, String encoding) throws IOException {
+        if (!Files.exists(filePath)) {
+            throw new IOException("File not found: " + filePath);
+        }
+        if (!Files.isRegularFile(filePath)) {
```
Comment (namrata111f) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:115]: Do we need to verify it is not a binary file or this much check is good enough? Also we are overwriting in the output file. Just to double check that is expected instead of appending.

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java [hunk 6]
```diff
@@ -0,0 +1,122 @@
...
+            // Get language
+            Language language = LanguageRegistry.PMD.getLanguageById(inputData.language);
+            if (language == null) {
+                throw new RuntimeException("Language not supported: " + inputData.language);
+            }
+
+            // Create TreeExportConfiguration
+            TreeExportConfiguration config = new TreeExportConfiguration();
+            config.setLanguage(language);
+            config.setFormat("xml"); // Always XML format for v1
```
Comment (nikhil-mittal-165) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:51]: will this come from input data after v1 ? asking if its a good idea to hardcode it now ?

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdWrapper.java [hunk 7]
```diff
@@ -152,4 +182,38 @@ private static void invokeRunCommand(String[] args) {
...
+        String argsInputFile = args[0];
+        String resultsOutputFile = args[1];
+
+        Gson gson = new Gson();
+
+        // Read input data
+        PmdAstDumpInputData inputData;
+        try (FileReader reader = new FileReader(argsInputFile)) {
+            inputData = gson.fromJson(reader, PmdAstDumpInputData.class);
+        } catch (Exception e) {
```
Comment (nikhil-mittal-165) [inline packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdWrapper.java:199]: should we throw entire exception stack trace ? or just the exception message should be fine ? we can also think of logging the exception stack

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java (hunk 1 above)
Comment (aruntyagiTutu) [reply packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:36]: We are intentionally using System.out.println here to maintain consistency
  with the rest of the codebase (see PmdRuleDescriber.java:99 and
  CpdRunner.java:55). The project uses slf4j-nop and relies on stdout for
  inter-process communication with the TypeScript side.
  
  am I missing something @namrata111f ?

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java (hunk 2 above)
Comment (aruntyagiTutu) [reply packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:57]: updated names

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java (hunk 3 above)
Comment (aruntyagiTutu) [reply packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:59]: - Inner try-finally (lines 59-74): Ensures System.out is always restored and  
  capturedPrintStream is closed, even if an exception occurs
  - Outer try-catch (lines 35-86): Handles any exceptions and stores them in
  results.error

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java (hunk 4 above)
Comment (aruntyagiTutu) [reply packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:80]: same as above

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java (hunk 5 above)
Comment (aruntyagiTutu) [reply packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:115]: The validation here is intentionally lightweight and avoids reading file
  content PMD's TreeExporter will handle content
  validation when it attempts to parse the file.

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java (hunk 6 above)
Comment (aruntyagiTutu) [reply packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:51]: in my understanding it will always be xml, but still keeping it open, just in case.

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdWrapper.java (hunk 7 above)
Comment (aruntyagiTutu) [reply packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdWrapper.java:199]: updated

**Review by nikhil-mittal-165 — APPROVED** [review]

File: packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java (hunk 1 above)
Comment (namrata111f) [reply packages/code-analyzer-pmd-engine/pmd-cpd-wrappers/src/main/java/com/salesforce/sfca/pmdwrapper/PmdAstDumper.java:36]: Understood, was not aware that inter-process communication relies on stdout. Just curious for debugging do we log these stdout on the  TypeScript side?

**Review by namrata111f — APPROVED** [review]
